src/ast/compiler.rs

```diff
@@ -40,6 +40,173 @@ mod jit;
+    if is_present_only {
+        pb.finish_with_message("仅展示，中间代码编译完成后退出");
+        return;
+    }
+    pb.finish_with_message("中间代码编译完成");
+
+    let ctx = Context::create();
+
+    let (llvmmod, files) = process_llvm_ir(db, docs, &ctx, op);
+
+    pl_link(llvmmod, files, out.clone(), op);
+}
+
+/// process_llvm_ir retrieves llvm IR from db and docs and optimizes it with the help of llvm optimizer.
+pub fn process_llvm_ir<'a>(
+    db: &dyn Db,
+    docs: MemDocsInput,
+    ctx: &'a Context,
+    op: Options,
+) -> (Module<'a>, Vec<PathBuf>) {
+    let mods = compile_dry::accumulated::<ModBuffer>(db, docs);
+    ensure_no_error(db, docs);
+
+    let total_steps = if op.jit { 2 } else { 3 };
+    let pb = ProgressBar::new(mods.len() as u64);
+    prepare_prgressbar(&pb, op, format!("{}[{:2}/{:2}]", TRUCK, 2, total_steps));
+
+    let mut set = FxHashSet::default();
+    let mut output_files = vec![];
+
+    let tm = crate::ast::builder::llvmbuilder::get_target_machine(op.optimization.to_llvm());
+    let llvmmod = ctx.create_module("main");
+
+    for m in mods {
+        pb.inc(1);
+        let m = m.path;
+        if set.contains(&m) {
+            continue;
+        }
+        set.insert(m.clone());
+        let o = m.with_extension("o");
+        // println!("{}", m.clone().to_str().unwrap());
+        let module = Module::parse_bitcode_from_path(m.clone(), ctx)
+            .unwrap_or_else(|_| panic!("parse {} failed", m.to_str().unwrap()));
+        pb.set_message(format!(
+            "正在优化模块 {} ",
+            module.get_name().to_str().unwrap().yellow()
+        ));
+        run_pass(&module, op.optimization.to_llvm());
+        pb.set_message(format!(
+            "正在生成模块 {} 的目标文件",
+            module.get_name().to_str().unwrap().yellow()
+        ));
+        module.verify().unwrap();
+        tm.write_to_file(&module, inkwell::targets::FileType::Object, &o)
+            .unwrap();
+        output_files.push(o);
+        _ = llvmmod.link_in_module(module);
+    }
+    pb.finish_with_message("中间代码优化完成");
+    (llvmmod, output_files)
+}
+
+pub fn pl_link(llvmmod: Module, oxbjs: Vec<PathBuf>, out: String, op: Options) {
+    llvmmod.verify().unwrap();
+    if op.genir {
+        let mut s = out.to_string();
+        s.push_str(".ll");
+        let llp = Path::new(&s[..]);
+        fs::write(llp, llvmmod.to_string()).unwrap();
+    }
+    let mut fo = out.to_string();
+    let mut out = out;
+
+    out.push_str(".bc");
+
+    let tm = crate::ast::builder::llvmbuilder::get_target_machine(op.optimization.to_llvm());
+    llvmmod.set_triple(&tm.get_triple());
+    llvmmod.set_data_layout(&tm.get_target_data().get_data_layout());
+
+    let total_steps = if op.jit { 2 } else { 3 };
+    let pb = ProgressBar::new(1);
+    prepare_prgressbar(&pb, op, format!("{}[{:2}/{:2}]", CLIP, 3, total_steps));
+
+    if op.jit {
+        llvmmod.write_bitcode_to_path(Path::new(&out));
+        pb.set_message("JIT模式将模块写入目标文件");
```

Comment:
Again, progress bar is meant to tell user the progress of compile rather than log infomations.